docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,378 @@
+as mentioned above, we'll need to support the manager being written in
+something other than rust. The definitions below are designed to be easy to
+serialize and used across the FFI - probably using protobufs, although JSON
+would also be an option if there was a good reason to prefer that.
+
+```rust
+// The main structure passed to a sync implementation. This needs to carry
+// all state necessary to sync.
+struct GlobalEngineState {
+    // note: no "declined" or "enabled" here - that's invisible to the engine.
+
+    // the result of info/configuration.
+    config: GlobalConfig,
+
+    // The token server token, used for authenticating with the storage servers.
+    storage_token: Vec<u8>, // (??) not clear on the type here, but whatever.
+
+    // Info about the collections managed by this engine. In most cases there
+    // will be exactly 1.
+    collections: HashMap<String, CollectionState>
+}
+
+// The state for a collection. Most engines will manage a single collection, but
+// there may be many.
+struct CollectionState {
+    // The info/collections response for this collection.
+    collection_info: CollectionInfo,
+
+    // The current syncID for the engine, as read from meta/global. Will never
+    // be None as the sync manager is responsbile for generating an ID if
+    // it doesn't exist.
+    // Engines are responsible for persisting this and taking action when it
+    // differs from last time.
+    sync_id: SyncGuid,
+
+    // The current version for the engine, as read from meta/global.
+    // Not clear that "u32" is the best option here - I guess it's OK so
+    // long as we treat an invalid u32 representation sanely.
+    // Note that persisting this may not be required if there's only one
+    // version supported, as the value can be compared against a hard-coded
+    // value in the engine.
+    version: u32,
+
+    // The keys to be used for this collection. This may be the default key (ie,
+    // used for all collections) or one specific to the collection, but the
+    // engine doesn't need to know that.
+    keys: KeyBundle,
+
+    // Commands the engine should act upon before syncing.
+    commands: Vec<Command>
+}
+
+// Used to support commands. Will either be a strongly typed enum, or a
+// weakly typed thing (such as {command: String, args: Vec<String>})
+// TBD.
+struct Command {
+    ...
+}
+```
+
+## APIs
+## Sync Manager
+
+```rust
+// Get the list of engines which are declined. There's a bit of confusion here
+// between an "engine" and a "collection" - eg, assuming history ends up using
+// 2 collections, the UI will want 1 entry.
+// Maybe we define this as a "collection" and the app should ignore collections
+// it doesn't understand?
+// Will return None if we haven't yet done a first sync so don't know. Will
+// return an empty vec if we have synced but no engines are declined.
+fn get_declined() -> Option<Vec<String>>;
```

Comment:
Just to clarify expected usage... Given that list of disabled engines may change at any time:
- applications are expected to query this whenever they need to display a list of enabled/disabled engines
- to perform a `sync all`, applications should call `sync_now(None)`, instead of iterating list of what collections they think are enabled (this list could have changed)

Right?

Reply:
My next version has changed this significantly - there's no separate API for enabling/disabling, or for determining what the current disabled states are. Instead, you specify state changes you desire before you sync, and at the end of the sync you get told what the current set is.